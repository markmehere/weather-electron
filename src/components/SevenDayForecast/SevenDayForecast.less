.firstRow, .secondRow {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
}

.secondRow {
  margin-top: 2vw;
}

.panel {
  width: 16vw;
  text-align: center;
  padding: 1vw;
}

.panel div {
  margin: auto;
}

.overall, .loading {
  height: 55vh;
}

.loading div {
    margin-top: 6vw;
}

.loading p {
  margin-top: 6vw;
  color: #777;
}
