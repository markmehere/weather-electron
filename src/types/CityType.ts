export type CityType = {
  value: number; /* woeid */
  label: string; /* city, country code (e.g. Melbourne, AU) */
};