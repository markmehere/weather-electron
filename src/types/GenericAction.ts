import { Action } from "redux";

export interface GenericAction<T> extends Action {
  payload: T;
} 
