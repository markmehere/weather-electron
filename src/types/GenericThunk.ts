import { ThunkAction } from "redux-thunk";
import { AppStore } from "../redux/reducers";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type GenericThunk = ThunkAction<any, AppStore, undefined, any>;
