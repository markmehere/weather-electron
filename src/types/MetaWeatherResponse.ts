import { WeatherKind } from "./WeatherKind";

export interface ConsolidatedWeatherResponse {
  id: string;
  weather_state_name: string;
  weather_state_abbr: WeatherKind;
  created: string;
  applicable_date: string;
  min_temp: number;
  max_temp: number;
  wind_speed: number;
  wind_direction: number;
  air_pressure: number;
  humidity: number;
  visibility: number;
  predictability: number;
}

export interface MetaWeatherResponse {
  consolidated_weather: ConsolidatedWeatherResponse[]
}