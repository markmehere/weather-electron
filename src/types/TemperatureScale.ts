export enum TemperatureScale {
  Celsius,
  Fahrenheit
}
