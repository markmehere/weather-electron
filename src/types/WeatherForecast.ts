import { WeatherKind } from "./WeatherKind";

export interface WeatherForecast {
  max: number;
  min: number;
  date: string;
  kind: WeatherKind;
}
