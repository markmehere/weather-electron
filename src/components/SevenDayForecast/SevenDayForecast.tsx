import React from "react";
import styles from "./SevenDayForecast.less";
import { WeatherSymbol } from "./WeatherSymbol/WeatherSymbol";
import { WeatherTemperature } from "./WeatherTemperature/WeatherTemperature";
import { TemperatureScale } from "../../types/TemperatureScale";
import { AppStore } from "../../redux/reducers";
import { connect } from "react-redux";
import { WeatherForecast } from "../../types/WeatherForecast";
import { getForecast, GetForecastActionCreator, fakeGetForecast } from "../../redux/actions/weather";
import { CityType } from "../../types/CityType";
import "react-loader-spinner/dist/loader/css/react-spinner-loader.css";
import Loader from "react-loader-spinner";

const FIRST_ROW_COUNT = 5;

export interface SevenDayForecastProps {

  // The seven-day forecast
  forecast?: WeatherForecast[];

  // The city we are forecasting
  city: CityType;

  // The temperature scale to use for the forecast
  scale: TemperatureScale;

  // The error from the state
  error?: string;

  // Get forecast action creator
  getForecast: GetForecastActionCreator;

  // Fake get forecast action creator
  fakeGetForecast: GetForecastActionCreator;

}

export class SevenDayForecastClass extends React.Component<SevenDayForecastProps> {

  componentDidMount() {
      this.props.getForecast(this.props.city);
  }

  renderPanel = (forecast: WeatherForecast, key: string | number) => (<div className={styles.panel} key={key}>
      <WeatherSymbol kind={forecast.kind} date={forecast.date} />
      <WeatherTemperature max={forecast.max} min={forecast.min} scale={this.props.scale} />
    </div>);

  renderLoader = () => <div className={styles.loading}>
    <Loader type="Bars" height={100} width={100} color="#2c4c7c" />
  </div>;

  renderError = () => <div className={styles.loading}>
    <p>{this.props.error}</p>
  </div>;

  render() {
    if (this.props.error) return this.renderError();
    if (!this.props.forecast) return this.renderLoader();
    const firstRow = this.props.forecast.slice(0, FIRST_ROW_COUNT);
    const secondRow = this.props.forecast.slice(FIRST_ROW_COUNT);
  
    return <div className={styles.overall}>
      <div className={styles.firstRow}>
        { firstRow.map(this.renderPanel) }
      </div>
      <div className={styles.secondRow}>
        { secondRow.map(this.renderPanel) }
      </div>
    </div>;
  }

}

const mapStateToProps = (state: AppStore) => ({
  forecast: state.weather.current?.forecast,
  scale: state.settings.scale,
  city: state.settings.city,
  error: state.weather.error
});

const mapDispatchToProps = {
  getForecast: getForecast,
  fakeGetForecast: fakeGetForecast
};

export default connect(mapStateToProps, mapDispatchToProps)(SevenDayForecastClass);
