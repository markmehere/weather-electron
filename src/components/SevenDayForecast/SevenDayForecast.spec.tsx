import React from 'react';
import { shallow } from 'enzyme';
import { WeatherKind } from '../../types/WeatherKind';
import { TemperatureScale } from '../../types/TemperatureScale';
import { SevenDayForecastClass } from './SevenDayForecast';

describe('<SevenDayForecast />', () => {

  const defaultProps = {
    forecast: [
      {
        max: 20.2,
        min: -2.2,
        date: '2020-05-09',
        kind: WeatherKind.Clear
      },
      {
        max: 18.2,
        min: 4.6,
        date: '2020-05-10',
        kind: WeatherKind.HeavyCloud
      },
    ],
    city: { value: 1105779, label: 'Sydney, AU' },
    scale: TemperatureScale.Celsius,
    getForecast: () => undefined as any,
    fakeGetForecast: () => undefined as any
  };
  
  const defaultPropsFaren = {
    ...defaultProps,
    forecast: [ ...defaultProps.forecast, { max: 30.2, min: 6.4, date: '2020-05-11', kind: WeatherKind.LightCloud } ],
    scale: TemperatureScale.Fahrenheit
  };

  const defaultPropsError = {
    ...defaultProps,
    error: 'Bad news'
  };

  const defaultPropsNoData = {
    ...defaultProps,
    forecast: null as any
  };

  it('matches snapshot with celsius scale', () => {
    const result = shallow(<SevenDayForecastClass {...defaultProps}  />);
    expect(result).toMatchSnapshot();
  });

  it('matches snapshot with farenheit scale', () => {
    const result = shallow(<SevenDayForecastClass {...defaultPropsFaren}  />);
    expect(result).toMatchSnapshot();
  });

  it('matches snapshot with error', () => {
    const result = shallow(<SevenDayForecastClass {...defaultPropsError}  />);
    expect(result).toMatchSnapshot();
  });

  it('matches snapshot no data', () => {
    const result = shallow(<SevenDayForecastClass {...defaultPropsNoData}  />);
    expect(result).toMatchSnapshot();
  });
});
