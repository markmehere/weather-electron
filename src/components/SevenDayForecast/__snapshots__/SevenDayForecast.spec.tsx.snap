// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SevenDayForecast /> matches snapshot no data 1`] = `
<div
  className="loading"
>
  <Loader
    className=""
    color="#2c4c7c"
    height={100}
    style={Object {}}
    timeout={0}
    type="Bars"
    visible={true}
    width={100}
  />
</div>
`;

exports[`<SevenDayForecast /> matches snapshot with celsius scale 1`] = `
<div
  className="overall"
>
  <div
    className="firstRow"
  >
    <div
      className="panel"
      key="0"
    >
      <WeatherSymbol
        date="2020-05-09"
        kind="c"
      />
      <WeatherTemperature
        max={20.2}
        min={-2.2}
        scale={0}
      />
    </div>
    <div
      className="panel"
      key="1"
    >
      <WeatherSymbol
        date="2020-05-10"
        kind="hc"
      />
      <WeatherTemperature
        max={18.2}
        min={4.6}
        scale={0}
      />
    </div>
  </div>
  <div
    className="secondRow"
  />
</div>
`;

exports[`<SevenDayForecast /> matches snapshot with error 1`] = `
<div
  className="loading"
>
  <p>
    Bad news
  </p>
</div>
`;

exports[`<SevenDayForecast /> matches snapshot with farenheit scale 1`] = `
<div
  className="overall"
>
  <div
    className="firstRow"
  >
    <div
      className="panel"
      key="0"
    >
      <WeatherSymbol
        date="2020-05-09"
        kind="c"
      />
      <WeatherTemperature
        max={20.2}
        min={-2.2}
        scale={1}
      />
    </div>
    <div
      className="panel"
      key="1"
    >
      <WeatherSymbol
        date="2020-05-10"
        kind="hc"
      />
      <WeatherTemperature
        max={18.2}
        min={4.6}
        scale={1}
      />
    </div>
    <div
      className="panel"
      key="2"
    >
      <WeatherSymbol
        date="2020-05-11"
        kind="lc"
      />
      <WeatherTemperature
        max={30.2}
        min={6.4}
        scale={1}
      />
    </div>
  </div>
  <div
    className="secondRow"
  />
</div>
`;
